cmake/DaemonFlags.cmake

```diff
@@ -185,26 +185,42 @@ else()
     set_c_cxx_flag("-ffast-math")
     set_c_cxx_flag("-fno-strict-aliasing")
 
-    # Set arch on x86 to SSE2 minimum and enable CMPXCHG16B
-    if (ARCH STREQUAL "i686")
-        set_c_cxx_flag("-m32")
-        set_c_cxx_flag("-msse2")
-        set_c_cxx_flag("-mtune=generic")
-        try_c_cxx_flag_werror(MFPMATH_SSE "-mfpmath=sse")
-    elseif (ARCH STREQUAL "amd64")
-        set_c_cxx_flag("-m64")
-        set_c_cxx_flag("-mtune=generic")
-        try_c_cxx_flag_werror(MCX16 "-mcx16")
+    if (ARCH STREQUAL "amd64")
+        set(GENERIC_ARCH "x86-64")
+    elseif (ARCH STREQUAL "i686")
+        set(GENERIC_ARCH "i686")
     elseif (ARCH STREQUAL "arm64")
-        set_c_cxx_flag("-march=armv8-a")
+        # Armv8-A minimum: Cortex-A50.
+        set(GENERIC_ARCH "armv8-a")
     elseif (ARCH STREQUAL "armhf")
-        set_c_cxx_flag("-march=armv7")
-        set_c_cxx_flag("-mfloat-abi=hard")
-        set_c_cxx_flag("-mfpu=neon")
+        # Armv7-A with VFP minimum: Cortex-A5.
+        # Hard float ABI (mainstream 32-bit ARM Linux distributions).
+        set(GENERIC_ARCH "armv7-a")
     else()
         message(FATAL_ERROR "Unsupported architecture ${ARCH}")
     endif()
 
+    option(USE_GENERIC_BUILD "Enforce generic -march and -mtune compiler options" ON)
+    if (USE_GENERIC_BUILD)
+        set_c_cxx_flag("-march=${GENERIC_ARCH}")
+        set_c_cxx_flag("-mtune=generic")
+    endif()
+
+    option(USE_EFFICIENT_HARDWARE "Enforce usage of hardware features like SSE, NEON, VFP, MCX16, etc." ON)
+    if (USE_EFFICIENT_HARDWARE)
```

Comment:
- USE_GENERIC_BUILD=ON USE_EFFICIENT_HARDWARE=OFF is a configuration to build for the absolute minimum specifications of the supported architectures
-  USE_GENERIC_BUILD=ON USE_EFFICIENT_HARDWARE=ON is the release build; support most machines of the supported architectures except some super old ones
- USE_GENERIC_BUILD=OFF USE_EFFICIENT_HARDWARE=ON probably doesn't make sense?

Do I have this right?

Reply:
> - USE_GENERIC_BUILD=ON USE_EFFICIENT_HARDWARE=OFF is a configuration to build for the absolute minimum specifications of the supported architectures
>  - USE_GENERIC_BUILD=ON USE_EFFICIENT_HARDWARE=ON is the release build; support most machines of the supported architectures except some super old ones

Right.

> - USE_GENERIC_BUILD=OFF USE_EFFICIENT_HARDWARE=ON probably doesn't make sense?

I would not say  “doesn't make sense”, but it's very niche yes. I believe it's still good to keep `USE_EFFICIENT_HARDWARE` enabled even if `USE_GENERIC_BUILD` isn't.

We may make `USE_EFFICIENT_HARDWARE` to only be processed (and then listed) if `USE_GENERIC_BUILD` is enabled, but I believe it's a bad idea: disabling `USE_GENERIC_BUILD` may produce some i386 build without SSE or some very old stuff if that's the current compiler default, and the user would have no clue about what's going on because the disabled option would even not be listed as disabled.

I can imagine people disabling `USE_GENERIC_BUILD` by wrongly believing this would magically produce an optimized build, keeping `USE_EFFICIENT_HARDWARE` (which sounds appealing) enabled would make sure such tweaker doesn't disable SSE or things like that without knowing.

Reply:
Yeah USE_EFFICIENT_HARDWARE sounds kind of like it should do `-march=native`. 

I agree that conditional options are often a bad idea due to hiding information.

Maybe it could be a tri-state option like USE_GENERIC_BUILD=full, USE_GENERIC_BUILD=modern, or USE_GENERIC_BUILD=none

Reply:
I also thought about defining the state by a word, but I like the fact cmake actually knows what is a bool, and that you can cycle bool values in CMake GUIs.

For a word-based state it would be possible to name the variable `BUILD_PROFILE` and provide the options:

- `release`: lowest hardware + things like SSE
- `lowend`: lowest hardware
- `custom`: enforces nothing

This would make possible other states like:

- `native`:  `-march=native`

But then I'm unhappy with the fact someone selecting `custom` will get SSE not being disabled, this is still hiding information, I prefer to require the user to explicitly disable things like SSE on purpose.

While the word-based state allows to define more states like the `native` one, it would still fail to provide a very basic option to enable advanced features like SSE, something a bool is very efficient at. Technically providing the equivalent of `USE_GENERIC_BUILD=OFF USE_EFFICIENT_HARDWARE=ON` to let the compiler selects the architecture our the user pick a custom architecture while making sure advanced features don't get disabled in the process would require to define yet another state word. In the end people would have to learn our custom states instead of learning about their compiler. We can expect people not doing release builds to learn what their compiler do.

In all cases I like the idea to provide the user the ability to make sure advanced features like SSE don't get disabled without knowing about it, that means without explicitly asking for it, something the bool does very well.

Defining a word-based state system may be reconsidered if there is a need for something like a ready-to-use `native` build and especially if we want it to be portable over systems and definable on other compilers like MSVC. Right now the only use I see would be distros like Gentoo that install packages by building them (and then making binary portability optional) or some service providers that would rent ready-to-use game servers and optimizing them to the maximum for their hardware. Those needs are niche if not existing yet, and I assume those would know how to add a compiler-specific `-march=native` option to their build script.

Reply:
OK that's a decent argument for having a bool for those features. We should just have a better name instead of USE_EFFICIENT_HARDWARE. How about DISABLE_MODERN_CPU_FEATURES?

Reply:
I like the `USE_` prefix, and the fact the release build would require both the features to be ON, which looks less confusing to me. I don't really like the _modern_ word, but mentioning `CPU` directly is a good idea.

Why not:

- `USE_CPU_GENERIC_ARCHITECTURE`
- `USE_CPU_RECOMMENDED_FEATURES`

I also thought about the `ADVANCED` qualifier but then it would be misleading as people may think it would select the latest advanced features like AVX. My concern is exactly that: I don't want to give the user some ability to configure things without letting know he may disable something we recommend.

Reply:
RECOMMENDED_FEATURES is better than my idea. LGTM with that